pkg/cloud/user_credentials.go

```diff
@@ -0,0 +1,242 @@
+	ResolveAccount(*Account) error
+	ResolveUser(*User) error
+	ResolveUserKeys(*User) error
+	GetUserWithKeys(*User) (bool, error)
+}
+
+// Domain contains specifications that identify a domain.
+type Domain struct {
+	Name string
+	Path string
+	ID   string
+}
+
+// Account contains specifications that identify an account.
+type Account struct {
+	Name   string
+	Domain Domain
+	ID     string
+}
+
+// User contains information uniquely identifying and scoping a user.
+type User struct {
+	ID        string
+	Name      string
+	APIKey    string
+	SecretKey string
+	Account
+}
+
+// ResolveDomain resolves a domain's information.
+func (c *client) ResolveDomain(domain *Domain) error {
+	// A domain can be specified by Id, Name, and or Path.
+	// Parse path and use it to set name if not present.
+	tokens := []string{}
+	if domain.Path != "" {
+		// Split path and get name.
+		tokens = strings.Split(domain.Path, domainDelimiter)
+		if domain.Name == "" {
+			domain.Name = tokens[len(tokens)-1]
+		}
+		// Ensure the path begins with ROOT.
+		if !strings.EqualFold(tokens[0], "root") {
+			tokens = append([]string{"ROOT"}, tokens...)
+		} else {
+			tokens[0] = "ROOT"
+			domain.Path = strings.Join(tokens, domainDelimiter)
+		}
+	}
+
+	// Set present search/list parameters.
+	p := c.cs.Domain.NewListDomainsParams()
+	p.SetListall(true)
+	setIfNotEmpty(domain.Name, p.SetName)
+	setIfNotEmpty(domain.ID, p.SetId)
+
+	// If path was provided also use level to search for domain.
+	if level := len(tokens) - 1; level >= 0 {
+		p.SetLevel(level)
+	}
+
+	resp, retErr := c.cs.Domain.ListDomains(p)
+	if retErr != nil {
+		return retErr
+	}
+
+	// If the Id was provided.
+	if domain.ID != "" {
+		if resp.Count != 1 {
+			return errors.Errorf("domain ID %s provided, expected exactly one domain, got %d", domain.ID, resp.Count)
+		}
+		if domain.Path != "" && !strings.EqualFold(resp.Domains[0].Path, domain.Path) {
+			return errors.Errorf("domain Path %s did not match domain ID %s", domain.Path, domain.ID)
+		}
+		domain.Path = resp.Domains[0].Path
+		domain.Name = resp.Domains[0].Name
+		return nil
+	}
+
+	// Consider the case where only the domain name is provided.
+	if domain.Path == "" && domain.Name != "" {
+		if resp.Count != 1 {
+			return errors.Errorf(
+				"only domain name: %s provided, expected exactly one domain, got %d", domain.Name, resp.Count)
+		}
+	}
+
+	// Finally, search for the domain by Path.
+	for _, possibleDomain := range resp.Domains {
+		if possibleDomain.Path == domain.Path {
+			domain.ID = possibleDomain.Id
+			return nil
+		}
+	}
+
+	return errors.Errorf("domain not found for domain path %s", domain.Path)
+}
+
+// ResolveAccount resolves an account's information.
+func (c *client) ResolveAccount(account *Account) error {
+	// Resolve domain prior to any account resolution activity.
+	if err := c.ResolveDomain(&account.Domain); err != nil {
+		return errors.Wrap(err, "error encountered when resolving domain details")
```

Comment:
```suggestion
		return errors.Wrapf(err, "error encountered when resolving domain details: %s", account.Domain)
```
For more informational error message.